### DISCUSSION TITLE
Change how print() prints strings

### DISCUSSION THREAD

**Discussion by miv391** [godotengine/godot-proposals#4431] [Scripting]
Currently `print()` prints strings without quotation marks. Because of that it can be impossible to see whether some value is a string or a number. For example these three different dicts are printed exactly the same:

```gdscript
print({"abc": 1, "foo": 2, "bar": 3})        # number values
print({"abc": "1", "foo": "2", "bar": "3"})  # string values
print({"abc": 1, "foo": "2, bar:3"})         # string value that looks like an another key and value pair
```

Output:
```
{abc:1, foo:2, bar:3}
{abc:1, foo:2, bar:3}
{abc:1, foo:2, bar:3}
```

(The keys may be in different order depending on the Godot version, the example output is from v4.0.alpha6.)

I suggest that quotation marks are printed around strings so that the output would be:

```
{abc:1, foo:2, bar:3}
{abc:"1", foo:"2", bar:"3"}
{abc:1, foo:"2, bar:3"}
```

**Comment by Mickeon** [comment]
Are you sure you mean ALL Strings? Your proposal would, somewhat undesirably, do this, too:
```GDScript
var a = "World"
print("Hello ", a , "!") # outputs "Hello ""World""!"
```

**Comment by miv391** [comment]
> Are you sure you mean ALL Strings?

That's very good point. I checked how printing works in Javascript and the same system might be good also in GDScript. Plain strings are printed without quotation marks, but when string is part of some structure, e.g. array or dict, quotation marks are printed. For example (Javascript code):

```js
console.log('hello')
console.log([1, 2, 3])
console.log(['1', '2', '3'])
console.log({ foo: 1, bar: 'abc' })
console.log('hello', 'world', { foo: 1, bar: 'abc' })
```

Output:
```
hello
[ 1, 2, 3 ]
[ '1', '2', '3' ]
{ foo: 1, bar: 'abc' }
hello world { foo: 1, bar: 'abc' }
```

**Comment by Calinou** [comment]
This is already being tracked in https://github.com/godotengine/godot-proposals/issues/2623.

**Reply by miv391** [reply]
How should this discussion handled now that the feature is implemented (in Godot 4 alpha 8)? Some kind of tag to the title?

**Reply by Calinou** [reply]
The proposal has a `status: implemented` label; I added it to the discussion as well.
